### PR TITLE
Feature pharmacy files

### PR DESCRIPTION
**Pull request by windwardrail**
Adds support for batch validation via a `Pharmacy.toml` configuration file.

### CODE REVIEW

File: src/bin/main.rs [hunk 1]
```diff
@@ -136,12 +172,51 @@ fn check(args: &ArgMatches) -> HowserResult<Option<ValidationProblem>> {
     }
 }
 
+fn process_pharmacy_file(args: &ArgMatches) -> HowserResult<Vec<ValidationProblem>> {
+    let mut report: Vec<ValidationProblem> = Vec::new();
+
+    if let Some(filename) = args.value_of("pharmacy") {
+        if let Ok(pharmacy_contents) = get_file_contents(filename) {
+            if let Some(specs) = pharmacy_contents.parse::<Value>()?["Specs"].as_table() {
```
Comment (ZackPierce) [inline src/bin/main.rs]: Seems like the lack of an `else` here means that an empty or malformed pharmacy file would result in a silent `Ok`

File: src/bin/main.rs [hunk 2]
```diff
@@ -136,12 +172,51 @@ fn check(args: &ArgMatches) -> HowserResult<Option<ValidationProblem>> {
...
+                for rx_file in specs.keys() {
+                    if let Some(doc_file) = specs.get(rx_file).map_or(None, |value| value.as_str())
+                    {
+                        let validate_args = vec!["howser", "validate", rx_file, doc_file];
+                        let app = make_app();
+                        let matches = app.get_matches_from(validate_args);
+                        if let Some(sub_m) = matches.subcommand_matches("validate") {
+                            let mut problems = validate(sub_m)?;
+                            if args.is_present("fail-early") && !problems.is_empty() {
+                                return Ok(problems);
```
Comment (ZackPierce) [inline src/bin/main.rs]: The nesting has gotten quite deep here, you may want to consider method chaining using the likes of `ok_or` and `and_then` and `or` to flatten the flow.

File: src/bin/main.rs [hunk 3]
```diff
@@ -61,42 +67,71 @@ fn make_app<'a, 'b>() -> App<'a, 'b> {
...
                 ),
         )
         .subcommand(
             SubCommand::with_name("validate")
-                .about("Validate a Markdown document against an Rx Prescription file.")
+                .about("Validates a Markdown document against an .rx Prescription file.")
+                .help_message("Prints help information.")
+                .setting(AppSettings::ArgRequiredElseHelp)
+                .arg(
+                    Arg::with_name("pharmacy")
```
Comment (ZackPierce) [inline src/bin/main.rs]: I'm not certain whether there are competing design requirements that lead to the current layout, but if it's not too much trouble, I would recommend the use of subcommands rather than an argument to represent the `pharmacy` feature.  

```
howser check README.rx
howser validate README.rx README.md

howser pharmacy check Pharmacy.toml
howser pharmacy validate Pharmacy.toml
```


Adding `--pharmacy` as an argument to `validate` and `check` results in treating one of two other parameters (the target file) to a `validate` call as optional, where previously it was simply required.

In other words, my preference as a downstream user is to have known requirements for each portion of the command tree, rather than having to cross-compare various optional arguments to see which are relevant for each use case 

(see also: fail-early, which could be scoped entirely to the pharmacy subcommands, but currently pollutes the top-level documentation)

**Comment by ZackPierce** [comment]
General comment, non-blocking - it seems like there are actually two distinct PRs worth of content here: the enhanced error messages and the pharmacy support. For future reference, it's usually better to do medium sized PRs than one large one.

File: src/bin/main.rs [hunk 4]
```diff
@@ -136,12 +172,51 @@ fn check(args: &ArgMatches) -> HowserResult<Option<ValidationProblem>> {
...
+    let mut report: Vec<ValidationProblem> = Vec::new();
+
+    if let Some(filename) = args.value_of("pharmacy") {
+        if let Ok(pharmacy_contents) = get_file_contents(filename) {
+            if let Some(specs) = pharmacy_contents.parse::<Value>()?["Specs"].as_table() {
+                for rx_file in specs.keys() {
+                    if let Some(doc_file) = specs.get(rx_file).map_or(None, |value| value.as_str())
+                    {
+                        let validate_args = vec!["howser", "validate", rx_file, doc_file];
+                        let app = make_app();
```
Comment (ZackPierce) [inline src/bin/main.rs]: Essentially making an external call to the application's CLI rather than directly using its internal functions after the code flow has progressed past the initial arg parsing is fragile and inadvisable. It seems to suggest that the internal surfaces of the codebase are inaccessible to the program itself by treating its own capabilities as a black box.

Please change this to a straight function call to the appropriate validation subroutine without looping back through the CLI.

File: src/bin/main.rs [hunk 5]
```diff
@@ -136,12 +172,51 @@ fn check(args: &ArgMatches) -> HowserResult<Option<ValidationProblem>> {
     }
 }
 
+fn process_pharmacy_file(args: &ArgMatches) -> HowserResult<Vec<ValidationProblem>> {
```
Comment (ZackPierce) [inline src/bin/main.rs]: I recommend reducing the spread of `ArgMatches` as much as reasonably possible, since it represents a squishy bag of strings and encourages redundant presence-checking and interpretation.  The most straightforward remediation is to do immediate local extraction and parsing of relevant parameters as part of the command-flow matching and pass those near-primitives to the "real work" sort of functions.  

A slightly more robust but time-consuming approach would be to parse the contents of the args entirely to some properly-typed structure (either manually or via `structopt`), and dole out subsections as needed.

This is not a blocking critique, just a recommendation.

File: src/bin/main.rs [hunk 6]
```diff
@@ -136,12 +172,51 @@ fn check(args: &ArgMatches) -> HowserResult<Option<ValidationProblem>> {
...
+    let mut report: Vec<ValidationProblem> = Vec::new();
+
+    if let Some(filename) = args.value_of("pharmacy") {
+        if let Ok(pharmacy_contents) = get_file_contents(filename) {
+            if let Some(specs) = pharmacy_contents.parse::<Value>()?["Specs"].as_table() {
+                for rx_file in specs.keys() {
+                    if let Some(doc_file) = specs.get(rx_file).map_or(None, |value| value.as_str())
```
Comment (ZackPierce) [inline src/bin/main.rs]: Rather than getting the iterating through the keys and then later getting the values, it seems like it would be less redundant/error-prone to instead iterate through the entries (key-value pairs).

```rust
for (key, value) in specs.iter() {
    if let Some(doc_file) = value.as_str() {
        ...
    }
}
```

Separately, it seems like there's no remediation or identification of a problem if the `value` is not a string.

File: src/bin/main.rs (hunk 3 above)
Comment (windwardrail) [reply src/bin/main.rs]: I think those are good ideas. I was following the existing design requirements, but I think they are flexible.

**Comment by windwardrail** [comment]
@ZackPierce, I think I've addressed all your comments. General code cleanup you suggested, better error processing and separation of concerns in the cli, subcommand interface changes, and an integration test suite have been addressed.

File: src/bin/main.rs [hunk 7]
```diff
@@ -175,40 +175,38 @@ fn check(args: &ArgMatches) -> HowserResult<Vec<ValidationProblem>> {
...
+    let prescription_pairs = specs
+        .as_table()
+        .ok_or(HowserError::RuntimeError(format!("Error parsing pharmacy file {}.", filename)))?;
+
+    for (rx_file, doc_file) in prescription_pairs {
+        let document =
+            doc_file
+                .as_str()
+                .ok_or(HowserError::RuntimeError(
+                    "The document corresponding to {} could not be parsed as a string.".to_string()))?;
```
Comment (ZackPierce) [inline src/bin/main.rs]: Without the use of `format!` and a corresponding arg, that `{}` is going to be included in the output string as-is.

File: requirements/software.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    on a per-project basis. Pharmacy files are written in toml and checked
+    into the repository along with the source code. They specify a series
+    of key value pairs that correspond to prescription files and the
+    documents that are to be checked with them.
+    
+    **Rx spec** = A DSL for constraining the content of Markdown 	documents
+
+    **prescription file (.rx)** =  A document written using the Rx spec DSL
+
+    **Pharmacy file (.toml)** = A list of documents to validate paired with the 	prescription they’re validated against
```
Comment (ZackPierce) [inline requirements/software.md]: Looks like there are some funny tabs/spacing included in this line and the `**Rx spec**` one above.

File: requirements/software.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    #### Background
+    
+    * **Given** that there exists some valid prescription files arbitrarily located within the project
+        * **And** the prescription files are divided into groups "matching" and "mis-matching"
+        * **And** for each of the prescription files in the "match" group there exists a corresponding markdown file somewhere in the project that matches the prescription
+        * **And** for each of the prescription files in the "mis-matching" group there exists a corresponding markdown file somewhere in the project that does not match the prescription
+        * **And** there exists a "pharmacy.toml" file 
+        * **And** the pharmacy file contains a toml section labeled "Specs"
+        * **And** for each prescription file there exists a toml key-value pair in the "Specs" section of the pharmacy file with filename as key and markdown filename as value
```
Comment (ZackPierce) [inline requirements/software.md]: To be more precise, this could be refined to "with Rx spec filename as key and markdown filename as value", otherwise it is ambiguous what filename is the key.

File: src/bin/main.rs [hunk 10]
```diff
@@ -50,31 +50,44 @@ fn run(args: &ArgMatches) -> HowserResult<()> {
...
+            match sub_m.subcommand() {
+                ("check", Some(sub_m)) => {
+                    let options = vec![CLIOption::VerboseMode(sub_m.is_present("verbose"))];
+                    let fail_early = sub_m.is_present("fail-early");
+                    let filename = sub_m
+                        .value_of("pharmacy").
+                        ok_or(HowserError::RuntimeError(
+                            "Pharmacy filename could not be parsed from the argument string.".to_string()))?;
+                    let pharmacy = parse_pharmacy_file(filename)?;
+                    let keys: Vec<_> = pharmacy.keys().collect();
```
Comment (ZackPierce) [inline src/bin/main.rs]: If you want to extend the notion of `pharmacy check` just a little bit further to parity with the single spec check variant, it would probably be to check the correctness of the pharmacy.toml file itself. E.G. confirming that all the key-value pairs have strings for their values.

This is a completeness suggestion, and not a blocking concern.

**Comment by ZackPierce** [comment]
Approved the changes. Saw a handful of minor issues in the new code/docs, but nothing critical.

**Comment by windwardrail** [comment]
@ZackPierce, updated to address latest comments.
